warns if duplicates in rename

wibeasley opened the pull request:
Related to #127 and #194

hadley commented on tests/testthat/test-rename.r:
```diff
@@ -37,3 +40,68 @@ test_that("Renaming lists", {
+  x <- list(a = 1, b = 2, c = 3)
+  replace_list <- c("c" = "f", "b" = "e", "a" = "f")
+  expected_response <- "The plyr::rename operation has created duplicates for the following name\\(s\\): \\(`f`\\)"
+  result <- rename(x=x, replace=replace_list, warn_duplicate = FALSE)
```
There doesn't seem to be an expectation here?

hadley commented on tests/testthat/test-rename.r:
```diff
@@ -1,4 +1,7 @@
-context("Rename")
+#################################################
```
Can you please use `// title ------` style?

wibeasley replied:
I'm sorry, I don't understand.  I don't see any examples like that in the package (even when I replace the forward slashes with pound signs).  Do you mean change
```r
#################################################
### Duplicate Names
#################################################
```

to
```r
## Duplicate Names -------------------
```
?

hadley replied:
Sorry, was in C++ mode. Yes - the standard for section headings is 

`# Section -----------------------------------`

hadley commented on R/rename.r:
```diff
@@ -16,7 +18,16 @@
+  
+  # Check if any names are duplicated.
+  duplicated_names <- names(x)[duplicated(names(x))]
+  if( warn_duplicate && (length(duplicated_names) > 0L)  ) {
```
Please follow the spacing guildelines at http://r-pkgs.had.co.nz/r.html#style

hadley commented on R/rename.r:
```diff
@@ -16,7 +18,16 @@
+  duplicated_names <- names(x)[duplicated(names(x))]
+  if( warn_duplicate && (length(duplicated_names) > 0L)  ) {
+    response_message <- paste0("The plyr::rename operation has created duplicates for the following name(s): (`", paste(duplicated_names, collapse="`, `"), "`)")
+    warning(response_message)
```
`warning()` automatically pastes together it's argument.

hadley commented on R/rename.r:
```diff
@@ -16,7 +18,16 @@
+rename <- function(x, replace, warn_missing = TRUE, warn_duplicate = TRUE ) {
+  
+  # This line does the real work of `rename()`.
   names(x) <- revalue(names(x), replace, warn_missing = warn_missing)
```
Shouldn't the checking code live in `revalue()` with the `warn_missing` checks?

wibeasley replied:
I don't feel strongly --you tell me.  

I believe there are many times where it's valid for `revalue()` to produce duplicates.  However, I think it's almost always a programming mistake for `rename()` to produce duplicates.  I don't remember coming across a scenario where it's desirable to have a dataset with two identical variable names.

Would you like me to move it to `revalue()` so the two functions have a consistent signature (even if it's rarely used in `revalue()`)?  I'll put it where you think it's best.

Since I'm already shaming my mother and compromising my integrating by conforming to your spacing and naming conventions....moving it to `revalue()` won't be so hard.

hadley commented on R/rename.r:
```diff
@@ -16,7 +18,16 @@
+    response_message <- paste0("The plyr::rename operation has created duplicates for the following name(s): (`", paste(duplicated_names, collapse="`, `"), "`)")
+    warning(response_message)
+  }  
+  return( x )
```
Please don't use explicit returns unless you're returning early.

hadley commented on R/rename.r:
```diff
@@ -12,11 +14,20 @@
 #' # Rename column "disp" to "displacement"
 #' rename(mtcars, c("disp" = "displacement"))
-rename <- function(x, replace, warn_missing = TRUE) {
+rename <- function(x, replace, warn_missing = TRUE, warn_duplicate = TRUE ) {
```
I think that should be `warn_duplicated` to match `warn_missing` ?

hadley commented on R/rename.r:
```diff
@@ -12,11 +14,20 @@
+  duplicated_names <- names(x)[duplicated(names(x))]
+  if (warn_duplicate && (length(duplicated_names) > 0L)) {
+    duplicated_names_message <- paste(duplicated_names, collapse="`, `")
+    warning("The plyr::rename operation has created duplicates for the following name(s): (`", duplicated_names_message, "`)")
```
Can you wrap those lines please? And all `call. = FALSE`.  

I think it's slightly more elegant to do `paste("`", duplicated_names, "`", collapse = ", ")`

hadley commented on tests/testthat/test-rename.r:
```diff
@@ -37,3 +38,61 @@ test_that("Renaming lists", {
+  x <- list(a = 1, b = 2, c = 3)
+  replace_list <- c("c" = "f", "b" = "e", "a" = "f")
+  expected_response <- "The plyr::rename operation has created duplicates for the following name\\(s\\): \\(`f`\\)"
+  expect_warning(object = rename(x = x, replace = replace_list), regexp = expected_response)
```
You don't need to name the arguments to expectations, and I'd shrink the match to "created duplicates"

hadley commented:
A few more minor style changes - getting close!

wibeasley commented:
I'm pretty sure I've addressed all your comments, except the idea of moving the part warning logic to `revalue()` (you can see those now hidden comments here: https://github.com/hadley/plyr/pull/234#discussion_r27511474).  Tell me if you want that moved to the function underlying `rename()`.

I'm not sure why my last commit failed Travis.  It looks like the test server had trouble finding the CRAN server.  I'm not authorized to restart the Travis build (I guess b/c I'm an outsider in the repo).

hadley commented:
Could you please merge/rebase?  @krlmlr can you please check the test coverage?

krlmlr commented:
I'll need to open a new pull request for that, based on this branch. @wibeasley: If you could sign up to http://coveralls.io, merge Hadley's master branch into yours and push, coverage should be reported automatically (although not as difference to current master). An alternative is running `covr::package_coverage()` locally.

krlmlr commented:
I have just pushed a copy of this branch to my repo, will post a link to the coveralls results.

wibeasley commented:
Ok.  Do you stil want me to do the pull request?

My changes affect only three files (the R, the Rd, and the test file).  Would it be easier if I started a new fork and pasted those three files into the fresh fork?

The last time I checked, there hadn't been any changes to those three files.  But I'll double check.

krlmlr commented:
I just checked coverage, didn't mean to take over ;-)

Coverage for your branch (after merging): https://coveralls.io/builds/2312893
Coverage for `master`: (will appear here)

Keeping this branch is probably simpler, but please make sure to delete `NEWS`.

wibeasley commented:
Some time I'll learn how to efficiently rebase and resolve something with 100+ commits.  I was faced with a lot of manual decisions, and I was worried about making the wrong ones.
